### PR TITLE
[SPARK-14281][TESTS] Fix java8-tests and simplify their build

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch fixes a compilation / build break in Spark's `java8-tests` and refactors their POM to simplify the build. See individual commit messages for more details.

### CODE REVIEW

File: external/java8-tests/pom.xml [hunk 1]
```diff
@@ -87,74 +82,26 @@
...
-          <encoding>UTF-8</encoding>
-          <maxmem>1024m</maxmem>
+          <compilerVersion>1.8</compilerVersion>
         </configuration>
       </plugin>
       <plugin>
-        <!-- disabled -->
         <groupId>net.alchim31.maven</groupId>
         <artifactId>scala-maven-plugin</artifactId>
-        <executions>
```
Comment (JoshRosen) [inline external/java8-tests/pom.xml]: AFAIK the idea behind this logic was to ensure that Java files would not be compiled by the Scala maven plugin, but disabling this has the side-effect of preventing this module's Scala test source file from being compiled and run in the Maven build.

Instead, I think it's okay to just modify the Scala Maven Plugin's `javacArgs` configuration.

File: pom.xml [hunk 2]
```diff
@@ -1922,6 +1922,7 @@
               <JAVA_HOME>${test.java.home}</JAVA_HOME>
             </environmentVariables>
             <systemProperties>
+              <log4j.configuration>file:src/test/resources/log4j.properties</log4j.configuration>
```
Comment (JoshRosen) [inline pom.xml]: This seems to be necessary for some unexplained reason; without this, the removal of the explicit surefire configuration in the `java8-tests` POM leads to tons of log spam.

File: pom.xml [hunk 3]
```diff
@@ -1937,6 +1938,14 @@
             <failIfNoTests>false</failIfNoTests>
             <excludedGroups>${test.exclude.tags}</excludedGroups>
           </configuration>
+          <executions>
```
Comment (JoshRosen) [inline pom.xml]: Without this change, `mvn -pl :java8-tests_2.11 test` wasn't triggering these tests.

File: pom.xml [hunk 4]
```diff
@@ -2345,27 +2355,12 @@
...
-                  <goal>test-jar</goal>
-                </goals>
-              </execution>
-            </executions>
-          </plugin>
-        </plugins>
-      </build>
-
+      <activation>
+        <jdk>[1.8,)</jdk>
```
Comment (JoshRosen) [inline pom.xml]: Here's where the JDK auto-detection is performed.

File: pom.xml [hunk 5]
```diff
@@ -2345,27 +2355,12 @@
 
     <profile>
       <id>java8-tests</id>
-      <build>
-        <plugins>
-          <!-- Needed for publishing test jars as it is needed by java8-tests -->
```
Comment (JoshRosen) [inline pom.xml]: We now publish test JARs by default, so this is redundant.

**Comment by JoshRosen** [comment]
I have updated the pull request builder Jenkins job to use a Java 8 JDK, so these tests should now be run.

It just occurred to me that I may want to configure the `dev/sparktestsupport/modules.py` file to make sure that these tests are automatically run if only a subset of tests would otherwise be run.

**Comment by vanzin** [comment]
Out of curiosity, isn't all this work mostly throw away if we're really dropping jdk7 support? At that point you could just move the source files to core/src/test, and get rid of all the build-related code.

**Comment by JoshRosen** [comment]
Yep, this would be irrelevant if we dropped JDK 7. However, I'm doing some other work which may have an impact on Java 8 lambda compatibility so I wanted to fix these tests to make my life a little easier.

**Comment by srowen** [comment]
+1. "Oops". Yeah we probably want to keep these tests in any event since they specifically test Java 8. Bigger question is just how far we want to go to update and streamline Java code and tests to use lambdas.

**Comment by JoshRosen** [comment]
Any objections to these build changes or is this now good to merge?

**Comment by vanzin** [comment]
Go ahead, it's a net improvement on the current state.

**Comment by JoshRosen** [comment]
Merging now; thanks everyone.

**Pull request closed without merging**